plugin/hpsa/hpsa.py

```diff
@@ -863,3 +880,106 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
+            ctrl_data = ctrl_all_show[ctrl_name]
+            if 'Battery/Capacitor Count' not in ctrl_data or \
+               ctrl_data['Battery/Capacitor Count'] == 0:
+                continue
+
+            sys_id = ctrl_data["Serial Number"]
+
+            battery_type = Battery.TYPE_UNKNOWN
+            battery_status = _hp_battery_status_to_lsm(ctrl_data)
+            relative_charge_level = Battery.CHARGE_LEVEL_UNKNOWN
+            absolute_charge_level = Battery.CHARGE_LEVEL_UNKNOWN
+
+            if ctrl_data["Cache Backup Power Source"] == "Capacitors":
+                battery_type = Battery.TYPE_CAPACITOR
+            elif ctrl_data["Cache Backup Power Source"] == "Batteries":
+                battery_type = Battery.TYPE_CHEMICAL
+            else:
+                raise LsmError(
+                    ErrorNumber.PLUGIN_BUG,
+                    "Got unknown battery type: %s for system %s" %
+                    (ctrl_data["Cache Backup Power Source:"], sys_id))
+
+            for counter in range(0, int(ctrl_data["Battery/Capacitor Count"])):
+
+                lsm_batteries.append(
+                    Battery(
+                        "%s_BAT_%d" % (sys_id, counter),
+                        "Battery %d of %s" % (counter, ctrl_name),
+                        battery_type, relative_charge_level,
+                        absolute_charge_level, battery_status, sys_id, None))
+
+        return lsm_batteries
+
+    @_handle_errors
+    def volume_cache_info(self, volume, flags=Client.FLAG_RSVD):
+        """
+        Depend on command:
+            hpssacli ctrl slot=0 show config detail
+        To disable both read and write cache:
+            ctrl slot=3 ld 1 modify aa=enable
+        To enable no battery write cache:
+            ctrl slot=0 modify nbwc=enable
+        To disable physical disk cache:
+            ctrl slot=5 modify dwc=enable
+        """
+        if not volume.plugin_data:
+            raise LsmError(
+                ErrorNumber.INVALID_ARGUMENT,
+                "Ilegal input volume argument: missing plugin_data property")
+
+        (ctrl_num, array_num, ld_num) = volume.plugin_data.split(":")
+        ctrl_data = self._sacli_exec(
+            ["ctrl", "slot=%s" % ctrl_num, "show", "config", "detail"]
+            ).values()[0]
+
+        write_cache_type = Cache.TYPE_UNKNOWN
+        read_cache_type = Cache.TYPE_UNKNOWN
+        disk_cache = Volume.DISK_CACHE_UNKNOWN
+
+        ld_info = ctrl_data.get(
+            "Array: %s" % array_num, {}).get("Logical Drive: %s" % ld_num, {})
+
+        if not ld_info:
+            raise LsmError(ErrorNumber.NOT_FOUND_VOLUME, "Volume not found")
+
+        flag_battery_ok = False
+
+        if 'Battery/Capacitor Count' in ctrl_data and \
+           int(ctrl_data['Battery/Capacitor Count']) >= 1 and \
+           ctrl_data['Battery/Capacitor Status'] == 'OK':
+            flag_battery_ok = True
+
+        # TODO(Gris Ge): support SSD cache(HP SmartCache).
```

Comment:
I see this comment here...I agree with you that it'd be great to support SmartCache in some way, but I don't necessarily think it should fold in with the BBWC implementation.